splitio/engine/unique_keys_tracker.py

```diff
@@ -0,0 +1,89 @@
+import abc
+import threading
+import logging
+from splitio.engine.filters.bloom_filter import BloomFilter
+
+_LOGGER = logging.getLogger(__name__)
+
+class BaseUniqueKeysTracker(object, metaclass=abc.ABCMeta):
+    """Unique Keys Tracker interface."""
+
+    @abc.abstractmethod
+    def track(self, key, feature_name):
+        """
+        Return a boolean flag
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def start(self):
+        """
+        No return value
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def stop(self):
+        """
+        No return value
+
+        """
+        pass
+
+class UniqueKeysTracker(BaseUniqueKeysTracker):
+    """Unique Keys Tracker class."""
+
+    def __init__(self, cache_size=30000, max_bulk_size=5000, task_refresh_rate = 24):
+        self._cache_size = cache_size
+        self._max_bulk_size = max_bulk_size
+        self._task_refresh_rate = task_refresh_rate
+        self._filter = BloomFilter(cache_size)
+        self._lock = threading.RLock()
+        self._cache = {}
+        # TODO: initialize impressions sender adapter and task referesh rate in next PR
+
+    def track(self, key, feature_name):
+        """
+        Return a boolean flag
+
+        """
+        if self._filter.contains(feature_name+key):
+            return False
+
+        with self._lock:
+            self._add_or_update(feature_name, key)
+            self._filter.add(feature_name+key)
+
+        if len(self._cache[feature_name]) == self._cache_size:
+            _LOGGER.warn("MTK Cache size for Split [%s] has reach maximum unique keys [%d], flushing data now.", feature_name, self._cache_size)
+#            TODO: Flush the data and reset split cache in next PR
+        if self._get_dict_size() >= self._max_bulk_size:
```

Comment:
you can control all the elements with an internal variable instead of calculating every time that you add an element